design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Boolean getBoolean(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a boolean.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Boolean getBoolean(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a decimal.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  BigDecimal getDecimal(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a decimal.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  BigDecimal getDecimal(String columnName);
+}
+```
+
+We considered representing column types in a more structured form (rather than plain strings) to accomodate complex/nested data types but felt this added complexity would not be helpful for most use cases. (This would also require a server-side change, so it's a fair bit of additional work.)
+
+For `getDecimal(...)` in the `Row` interface, rather than trying to parse the column type to extract the precision and scale, we will simply convert the value to a BigDecimal without explicitly specifying the precision and scale. We could also add an option for users to specify the scale and precision in the getter, if we think that would be useful.
+
+We could also add `getList(...)` and `getMap(...)` methods to the `Row` interface, but it's not clear to me how valuable this would be given that the nested data types would not be known.
+
+Tim also proposed switching from generic Java types (`Map<String, Object>`, `List<Object>`) to Vert.x types (`JsonObject`, `JsonArray`) as the latter is more type-safe and the client already has a dependency on Vert.x anyway. The downside, though, is that then apps that use the client would be required to depend on Vert.x as well.
+
+### Transient queries -- Non-streaming
+
+The `Client` interface will also provide the following methods for receiving the results of a transient query (push or pull) in a single batch (non-streaming),
+once the query has completed:
+```
+  /**
+   * Execute a query (push or pull) and receive all result rows together, once the query has
+   * completed.
+   *
+   * @param sql statement of query to execute.
+   * @return query result.
+   */
+  CompletableFuture<List<Row>> executeQuery(String sql);
+
+  /**
+   * Execute a query (push or pull) and receive all result rows together, once the query has
+   * completed.
+   *
+   * @param sql statement of query to execute.
+   * @param properties query properties.
+   * @return query result.
+   */
+  CompletableFuture<List<Row>> executeQuery(String sql, Map<String, Object> properties);
+```
+
+For a query to "complete" could mean:
+* The query is a pull query
+* The query is a push query with a limit clause, and the limit has been reached
+* The query is a push query that has been terminated
```

Comment:
maybe i'm missing something, but how do you terminate a push query using the client if you're waiting on it?

Reply:
The client is async. Say you issue a push query via `streamQuery(...)`. That method will return and you can stream results from the `QueryResult` at your leisure, then call `terminatePushQuery(...)` when you're done with it.

Reply:
Sorry, I specifically mean in the non-streaming case for push queries (that's the case I don't understand)
```
   CompletableFuture<List<Row>> executeQuery(String sql, Map<String, Object> properties); 
```
You don't get back a `QueryResult` in that case

Reply:
Returning the CompletableFuture from `executeQuery(...)` is async. If the user chooses to block on waiting for the CompletableFuture to complete, then you're right they won't be able to issue `terminatePushQuery(...)` during that time, but that's their decision then. Rather than blocking, they could check if the CompletableFuture has completed or not and choose to terminate the query after some time if they wish.

Reply:
I still don't understand, I'll walk you through how I would use this and then see if I'm missing something:
```java
CompletableFuture<List<Row>> result = executeQuery("SELECT * FROM foo EMIT CHANGES;");

// wait for 100ms, which is the refresh rate on my app
Thread.sleep(100);

if (!result.isDone()) {
  // I want to terminate the query before the next line, how do I do that?
}

// return whatever I was able to fetch in 100ms
return result.get();
```

Reply:
Oh, is your question just about how to get the query ID? I mistakenly thought it was about blocking/waiting on the CompletableFuture. We can update the return type from `List<Row>` to something that also includes the query ID, though that certainly makes the interface more complicated. Hmm...

Reply:
If you had the query ID you'd just replace the line where the comment is with `terminatePushQuery(queryId);`

Reply:
It this really a "valid" usage pattern? I would strictly distinguish push and pull query result as both are _semantically_ different and avoid the question all together... Even if a push query may return a finite result, I would never return it as a result _set_

Reply:
If you know a push query is going to be limited to 10 rows, then it's perfectly valid to return it as a List of Row. E.g. you might want the next 10 changes that happen to this table. Otherwise you'd have to subscribe to the publisher yourself, and manually collect the rows in a list. It's a time saver.

Reply:
> Otherwise you'd have to subscribe to the publisher yourself, and manually collect the rows in a list. It's a time saver.

Well, seem rather straightforward to do this if needed. I prefer to add syntactic sugar not in the first version of an API, but only after (frequent) user request, and stick to a "clean" separation of concern first. It's pretty hard to judge upfront if this pattern is common or not (my gut feeling is, it's not common -- of course, I don't have any data to backup this claim) so it might be better to be conservative.

Reply:
>> Well, seem rather straightforward to do this if needed.

For some definition of "straightforward" I guess ;)

You'd have to create a class (the subscriber). Subscribe to it, manually collect the rows, deal with completable futures, etc. Not really that simple for a first time user.

Reply:
Fair enough. I just see the "danger" that people use `poll()` on a query that return an infinite result. Or would we not allow this and throw an error upfront (might be easy to detect this case)?

Reply:
What's wrong with calling poll() on a non terminating query? Whether use the sync or async api is orthogonal to whether the query is terminating or not.